### ISSUE TITLE
Workflows configured to run on a CRON schedule stop after a few executions

### ISSUE THREAD

**Issue by arvindsampath** [temporalio/temporal#424]
## Expected Behavior
Workflows configured to run on a CRON schedule run at the configured frequency until the temporal service is shut down if they are not cancelled

## Actual Behavior
CRON workflows stop executing after a few times. This issue is observed in the server release version 0.23.1. There are some WARN level logs in the service console with the below message

> "msg":"New timer generated is less than read level"

They work well in the server version 0.21.1 and Java SDK 0.19.0.

## Steps to Reproduce the Problem

1. Start a workflow with the CRON Pattern "*/2 * * * *" to execute the workflow every 2 minutes
 

## Specifications

  - Version: 0.23.1
  - Platform: Java SDK, Service runs inside a docker container

**Comment by samarabbas** [comment]
Can you share the history of last execution of workflow?  

Make sure your workflow timeouts are not very aggressive.  Let's say your workflow timeout is 5 minute and Temporal server is down for more than 5 minutes than it may timeout the workflow instead of firing off cron run.

**Comment by arvindsampath** [comment]
I will run new tests and share the artifacts. In my case, the workflow wasn't doing much and returning right away. We had changed the various options timeout options to match the new SDK library APIs. I will double check them and post here.

**Comment by arvindsampath** [comment]
Thanks for the tip @samarabbas. I revisited the timeout options and was able to get past the issue. In the previous version of the SDK, we had used "setExecutionStartToCloseTimeout()". The signatures changed a bit with the new release and we now the options "setWorkflowRunTimeout()" and "setWorkflowExecutionTimeout()". We updated the option from "setExecutionStartToCloseTimeout()" to  "setWorkflowExecutionTimeout()". I think this is causing the CRON to timeout after the specified duration. I was able to get past the issue by using "setWorkflowRunTimeout()".

SDK version 0.19.0 (works)
`WorkflowOptions.newBuilder()
                .setExecutionStartToCloseTimeout(Duration.ofMinutes(6))
                .setTaskStartToCloseTimeout(Duration.ofMinutes(1))`

**SDK version 0.19.0 (works)**
`WorkflowOptions.newBuilder()
                .setExecutionStartToCloseTimeout(Duration.ofMinutes(6))
                .setTaskStartToCloseTimeout(Duration.ofMinutes(1))`

**SDK version 0.23.1 (CRON workflow stops executing after couple runs)**
`WorkflowOptions.newBuilder()
                .setWorkflowExecutionTimeout(Duration.ofMinutes(6))
                .setWorkflowTaskTimeout(Duration.ofMinutes(1))`

**SDK version 0.23.1 (CRON workflow runs as expected)**
`WorkflowOptions.newBuilder()
                .setWorkflowRunTimeout(Duration.ofMinutes(6))
                .setWorkflowTaskTimeout(Duration.ofMinutes(1))`

In summary, "WorkflowRunTimeout" is the timeout for each workflow run whereas "WorkflowExecutionTimeout" is the timeout for the overall execution and in the case of CRON workflows, since the same workflow ID is used for each periodic execution, this parameter caps the total duration till which the workflow schedule is valid.The previous version of the SDK had just one timeout parameter and that was applied to each periodic run.

**Comment by samarabbas** [comment]
Cool.  WorkflowExecutionTimeout is applicable across all runs of an execution.  Set WorkflowRunTimeout for cron use cases which runs for ever and do not set WorkflowExecutionTimeout.
